python/shotgun_model/shotgun_entity_model.py

```diff
@@ -0,0 +1,258 @@
+            return sg_data
+
+        field_data = get_sanitized_data(item, self.SG_ASSOCIATED_FIELD_ROLE)
+        field_value = field_data.get("value")
+        if (field_value 
+            and isinstance(field_value, dict) 
+            and "id" in field_value 
+            and "type" in field_value):
+            return field_value
+
+        return None
+
+    def async_refresh(self):
+        """
+        Trigger an asynchronous refresh of the model
+        """
+        self._refresh_data()
+
+    def _populate_default_thumbnail(self, item):
+        """
+        Whenever an item is constructed, this methods is called. It allows subclasses to intercept
+        the construction of a QStandardItem and add additional metadata or make other changes
+        that may be useful. Nothing needs to be returned.
+
+        :param item: QStandardItem that is about to be added to the model. This has been primed
+                     with the standard settings that the ShotgunModel handles.
+        :param sg_data: Shotgun data dictionary that was received from Shotgun given the fields
+                        and other settings specified in load_data()
+        """
+        found_icon = False
+
+        # get the associated field data with this node
+        field_data = get_sanitized_data(item, self.SG_ASSOCIATED_FIELD_ROLE)
+        # get the full sg data for this node (leafs only)
+        sg_data = get_sg_data(item)
+
+        # {'name': 'sg_sequence', 'value': {'type': 'Sequence', 'id': 11, 'name': 'bunny_080'}}
+        field_value = field_data["value"]
+
+        if isinstance(field_value, dict) and "name" in field_value and "type" in field_value:
+            # this is an intermediate node which is an entity type link
+            entity_icon = self._get_default_thumbnail(field_value)
+            if entity_icon:
+                # use sg icon!
+                item.setIcon(QtGui.QIcon(entity_icon))
+                found_icon = True
+
+        elif sg_data:
+            # this is a leaf node!
+            entity_icon = self._get_default_thumbnail(sg_data)
+            if entity_icon:
+                # use sg icon!
+                item.setIcon(QtGui.QIcon(entity_icon))
+                found_icon = True
+
+        # for all items where we didn't find the icon, fall back onto the default
+        if not found_icon:
+            item.setIcon(QtGui.QIcon(self._default_icon))
+
+    def _get_default_thumbnail(self, sg_entity):
+        """
+        Get the default icon for the specified entity.
+
+        :param sg_entity:   A Shotgun entity dictionary for the entity to get the
+                            icon for.
+        :returns:           A QIcon for the entity if available.  For Step entities, a
+                            swatch representing the step colour is returned.  If no
+                            icon is available for the entity type then None is returned
+        """
+        if sg_entity.get("type") == "Step":
```

Comment:
nah, I think that would be over specialisation - the model is supposed to work for any entity type but it just happens that the icon for Step is dynamic (a swatch).  You may also use this model for another entity type but include Step somewhere in the hierarchy so the icon would be needed there anyway.